docs/uenv-compilation-spack.md

```diff
@@ -0,0 +1,163 @@
+
+!!! note
+    While this guide tries to explain everything step-by-step, it might be difficult to follow without any knowledge of [Spack]. Please have a look at [Spack Basic Usage] for a short introduction to [Spack].
+
+The uenv of a supported application contains all the dependencies to build the software with a particular configuration. In [Spack], such configuration is defined by a [spec] (see [Spack Basic Usage] for more details). Most uenvs already provide modules or [Spack Filesystem Views] which allow to manually build the same configuration. However, it is possible that you want to build or develop an application with a different configuration. To avoid re-building the whole uenv, you can re-use what is already there and build your new configuration using [Spack]. 
+
+This guide explains a _developer workflow_ allowing to either build your own package with [Spack], or use [Spack] to build all the dependencies and provide a build environment for the package.
+
+!!! note
+    This guide assumes that you have a local installation of [Spack]. If you don't have [Spack] installed, follow [Spack Getting Started].
+
+!!! warning
+    Avoid installing [Spack] on `HOME`. Packages are installaed within the `spack/` folder, and you might quickly run out of space.
+
+## Example: CP2K
+
+As an example, we will consider [CP2K] as the application we want to develop. While the uenv for [CP2K] we provide is rather complete, let's assume for simplicity that the provided configuration is very bare bone:
+
+```
+cp2k@2024.1 +cuda cuda_arch=80
+```
+
+This [spec] defines a build of [CP2K] version `2024.1` with CUDA acceleration. All other dependencies and features are defined by the default values in the [CP2K Spack package].
+
+## Set uenv as upstream Spack instance
+
+Activate the uenv. Here we assume the uenv described above is called `cp2k/2024.1` and it is already deployed:
+
+```bash
+uenv image pull cp2k/2024.1
+uenv start cp2k/2024.1
+```
+
+With the uenv active, we can now tell our local [Spack] instance to use the uenv as an upstream [Spack] instance (see [Chaining Spack Installations] for more details):
+
+```bash
+export SPACK_SYSTEM_CONFIG_PATH=/user-environment/config/
+```
+
+!!! note
+    We assumed here that the uenv is mounted in the standard location `/user-environment/`. If it is mounted in a non-standard location, adjust the previous command accordingly.
+
+## Building your own version 
+
+Let's assume we want to have a version of [CP2K] what uses the COSMA library for communication-optimal matrix multiplication with GPU-aware MPI:
+
+```
+cp2k@2024.1 +cuda cuda_arch=80 +cosma ^cosma +gpu_direct
+```
+
+COSMA is not availabe in the uenv described above since it is not a dependency needed by `cp2k@2024.1 +cuda cuda_arch=80`. 
```

Comment:
It might be worth mentioning here that in general it is `cuda_arch=90`, otherwise people might copy/paste the wrong arch.

Reply:
Yeah, I realized too late this is an annoying one... `cp2k@2024.1` can't be build with `cuda_arch=90`. Maybe worth using a different example (like SIRIUS) where we can actually use `cuda_arch=90`?

Reply:
Yes, or QuantumESPRESSO%nvhpc.

Or adding `packages:all:prefer:cuda_arch=90`, but then set it explicitly to 80 in the spec and explain why.

Reply:
For QE I've added the spack upstream example in the PR: https://github.com/eth-cscs/alps-uenv/pull/92
But it's not as well explained as you did.

Reply:
Thanks for pointing it out, I missed it. Having it in a central location would avoid repetition, since it's useful for many uenvs. I'll look at what you wrote already, and incorporate it with this PR. I'll switch package too, to avoid confusion.

> Or adding packages:all:prefer:cuda_arch=90, but then set it explicitly to 80 in the spec and explain why.

This is what I did for the recipe, but is an additional layer of information that a general user does not need to know about (it's CP2K-specific).